controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -345,130 +364,268 @@ public String getScopeName() {
-                                                                segmentIndex.getData(),
-                                                                segmentTable.getData(),
-                                                                truncationRecord))
-                                        ))));
+    public CompletableFuture<Map<Segment, Long>> getSegmentsAtHead() {
+        // read current epoch record
+        return getTruncationRecord()
+                .thenCompose(truncationRecord -> {
+                    if (truncationRecord.getObject().equals(StreamTruncationRecord.EMPTY)) {
+                        return getSegmentsInEpoch(0)
+                                .thenApply(segments -> segments.stream().collect(Collectors.toMap(x -> x, x ->  0L)));
+                    } else {
+                        return CompletableFuture.completedFuture(truncationRecord.getObject().getStreamCut().entrySet()
+                                                                                 .stream().collect(Collectors.toMap(x ->
+                                                transform(truncationRecord.getObject().getSpan().keySet().stream()
+                                                                          .filter(y -> y.segmentId() == x.getKey()).findFirst().get()),
+                                        Map.Entry::getValue)));
+                    }
+                });
     }
 
     @Override
-    public CompletableFuture<List<Long>> getActiveSegments(final int epoch) {
-        return getHistoryIndex()
-                .thenCompose(historyIndex -> getHistoryTable()
-                        .thenApply(historyTable -> TableHelper.getSegmentsInEpoch(historyIndex.getData(), historyTable.getData(), epoch)));
+    public CompletableFuture<List<Segment>> getSegmentsInEpoch(final int epoch) {
+        return getEpochRecord(epoch)
+                .thenApply(epochRecord -> transform(epochRecord.getSegments()));
     }
 
     @Override
     public CompletableFuture<List<Segment>> getSegmentsBetweenStreamCuts(Map<Long, Long> from, Map<Long, Long> to) {
-        return getHistoryIndex()
-                .thenCompose(historyIndex -> getHistoryTable()
-                        .thenCompose(historyTable -> getSegmentIndex()
-                                .thenCompose(segmentIndex -> getSegmentTable()
-                                        .thenApply(segmentTable ->
-                                                TableHelper.findSegmentsBetweenStreamCuts(historyIndex.getData(), historyTable.getData(),
-                                                        segmentIndex.getData(), segmentTable.getData(), from, to)))));
+        return segmentsBetweenStreamCuts(from, to).thenApply(x -> new LinkedList<>(transform(x)));
+    }
+
+    private CompletableFuture<Set<StreamSegmentRecord>> segmentsBetweenStreamCuts(Map<Long, Long> from, Map<Long, Long> to) {
+        // compute stream cut span for `from` till `to`
```

Comment:
I'm wondering how we are unit testing all these private methods. Their processing is fairly critical and we should consider having tests that exercise them individually to check for correct behavior.

Reply:
i have added unit test for this in StreamTestBase.java `testStreamCutSegmentsBetween`

Reply:
Where is it? I can't find the class or test.

Reply:
https://github.com/pravega/pravega/pull/3033/files#r228987671

Reply:
Right, this private method is being exercised via `getSegmentsBetweenStreamCuts`, which is a public method. What about the other private methods that follow this one, e.g., are they all being properly unit-tested?

Reply:
Most of these methods are covered through existing tests built against StreamMetadataStore interface.

For example: for sizeBetweenStreamCuts:
`StreamMetadataStoreTests.sizeTest()`
`StreamMetadataTasksTest.sizeBasedRetentionTest()`
`StreamTestBase.truncationTest()`